pipeline/authn/authenticator_oauth2_introspection.go

```diff
@@ -231,12 +260,7 @@ func (a *AuthenticatorOAuth2Introspection) Config(config json.RawMessage) (*Auth
 	var rt http.RoundTripper
 
 	if c.PreAuth != nil && c.PreAuth.Enabled {
-		rt = (&clientcredentials.Config{
-			ClientID:     c.PreAuth.ClientID,
-			ClientSecret: c.PreAuth.ClientSecret,
-			Scopes:       c.PreAuth.Scope,
-			TokenURL:     c.PreAuth.TokenURL,
-		}).Client(context.Background()).Transport
+		rt = a.preAuthCache.Client(context.Background(), c.PreAuth).Transport
```

Comment:
`c.PreAuth` is a parameter which may change per route, but the authenticator is instantiated only once with one global cache, which will cause errors when more than one PreAuth are configured. Instead, this should probably be cached using the c.PreAuth contents as the key (e.g. crc32ing a string representation of c.PreAuth) .

Another issue is that the cache is never invalidated, even if requests start failing. This is problematic when, for example, the configuration is changed and subsequently reloaded.